src/controller/cmcd-controller.ts

```diff
@@ -343,6 +353,62 @@ export default class CMCDController implements ComponentAPI {
     return info.len * 1000;
   }
 
+  /**
+   * Create a playlist loader
+   *
+   * @returns {PlaylistLoaderConstructor | undefined}
+   */
+  private createPlaylistLoader(): PlaylistLoaderConstructor | undefined {
+    const { pLoader } = this.config;
+
+    if (!this.config.cmcdEnabled) {
+      return pLoader;
+    }
+
+    const PlayListLoader =
+      pLoader || (this.config.loader as PlaylistLoaderConstructor);
+    const apply = this.applyPlaylistData;
+
+    return class CMCDPlaylistLoader extends PlayListLoader {
+      load(
+        context: PlaylistLoaderContext,
+        config: LoaderConfiguration,
+        callbacks: LoaderCallbacks<PlaylistLoaderContext>
+      ) {
+        apply(context);
+        super.load(context, config, callbacks);
+      }
+    };
+  }
+
+  /**
+   * Create a playlist loader
+   *
+   * @returns {FragmentLoaderConstructor | undefined}
+   */
+  private createFragmentLoader(): FragmentLoaderConstructor | undefined {
+    const { fLoader } = this.config;
+
+    if (!this.config.cmcdEnabled) {
+      return fLoader;
+    }
+
+    const FragmentLoader =
+      fLoader || (this.config.loader as FragmentLoaderConstructor);
+    const apply = this.applyFragmentData;
+
+    return class CMCDFragmentLoader extends FragmentLoader {
+      load(
+        context: FragmentLoaderContext,
+        config: LoaderConfiguration,
+        callbacks: LoaderCallbacks<FragmentLoaderContext>
+      ) {
+        apply(context);
+        super.load(context, config, callbacks);
+      }
+    };
```

Comment:
This is the main part of the refactor. When the `CMCDController` is instantiated, it will compose new loaders based on the `HlsConfig` using these functions.